internal/sharedcomponent/sharedcomponent.go

```diff
@@ -85,20 +74,72 @@ func (c *Component[V]) Unwrap() V {
 
 // Start starts the underlying component if it never started before.
 func (c *Component[V]) Start(ctx context.Context, host component.Host) error {
+	if c.hostWrapper == nil {
+		c.hostWrapper = &hostWrapper{
+			host:           host,
+			sources:        make([]componentstatus.StatusReporter, 0),
+			previousEvents: make([]*componentstatus.StatusEvent, 0),
+		}
+	}
+
+	statusReporter, isStatusReporter := host.(componentstatus.StatusReporter)
+	if isStatusReporter {
+		c.hostWrapper.addSource(statusReporter)
+	}
```

Comment:
Could all this be moved into `startOnce`? If so, I think it can be:

```suggestion
	if reporter, ok := host.(componentstatus.Reporter); ok {
		c.hostWrapper = &hostWrapper{
			host:           host,
			sources:        []componentstatus.Reporter{ reporter },
			previousEvents: make([]*componentstatus.Event, 0),
		}
	}
```

Reply:
We can't move `c.hostWrapper.addSource(statusReporter)` into `startOnce` since that is the function that is building our chain of sources. For simplicity when calling `c.component.Start` I believe we always want to use `hostWrapper`, even when `component.Host` does not implement `componentstatus.Reporter`, so we shouldn't move the construction of `hostWrapper` behind a condition that requires `componentstatus.Reporter`.

Reply:
Removing the condition makes sense if we're always instantiating the `hostWrapper`, but I'm not sure I understand why it would ever need to be instantiated more or less than once.

Reply:
I've tried reorganizing a little, does it seem better?

Reply:
You're correct, but I included the nil check to act as a way to not call `c.hostWrapper.addSource(statusReporter)` a second time, since now it needs to be called within `startOnce`